File: CortexSuite/vision/benchmarks/localization/src/matlab/quatRot.m
function ret=quatRot(vec, rQuat)
nr=size(vec,1);
tv = zeros(nr,1);
vQuat=[tv, vec];
temp = quatMul(rQuat, vQuat);
temp1 = quatConj(rQuat);
retVec = quatMul(temp, temp1);
%retVec=quatMul(quatMul(rQuat, vQuat),quatConj(rQuat));
%ret=retVec(:,2:4);
rows = size(retVec, 1);
ret = zeros(rows,3);

for i=1:rows
    k =1;
    for j=2:4
        ret(i,k) = retVec(i,j);
        k = k+1;
    end
end
%size(ret)

File: CortexSuite/vision/benchmarks/localization/src/c/quatRot.c
/********************************
Author: Sravanthi Kota Venkata
********************************/

#include <stdio.h>
#include <stdlib.h>
#include "localization.h"

F2D* quatRot(F2D* vec, F2D* rQuat)
{
    F2D *ret;
    int nr, i, j, k, rows, cols;
    F2D *tv, *vQuat, *temp, *temp1;
    F2D *retVec;

    nr = vec->height;
    tv = fSetArray(nr, 1, 0);
    vQuat = fHorzcat(tv, vec);
    temp = quatMul(rQuat, vQuat);
    temp1 = quatConj(rQuat);
    retVec = quatMul(temp, temp1);

    rows = retVec->height;
    cols = retVec->width;

    ret = fSetArray(rows, 3, 0);

    for(i=0; i<rows; i++)
    {
        k = 0;
        for(j=1; j<4; j++)
        {
            subsref(ret,i,k) = subsref(retVec,i,j);
            k++;
        }
    }

    fFreeHandle(tv);
    fFreeHandle(vQuat);
    fFreeHandle(temp);
    fFreeHandle(temp1);
    fFreeHandle(retVec);
        
    return ret;
}


